src/Microsoft.AspNetCore.Server.IISIntegration/Server/IISHttpServer.cs

```diff
@@ -55,15 +60,50 @@ public Task StartAsync<TContext>(IHttpApplication<TContext> application, Cancell
 
         public Task StopAsync(CancellationToken cancellationToken)
         {
-            // TODO: Drain pending requests
+            // First call back into native saying "DON'T SEND ME ANY MORE REQUESTS"
+            void RegisterCancelation()
+            {
+                cancellationToken.Register(() =>
+                {
+                    _shutdownSignal.TrySetResult(null);
+                });
+            }
+            if (Interlocked.Exchange(ref _stopping, 1) == 1)
+            {
+                RegisterCancelation();
 
-            // Stop all further calls back into managed code by unhooking the callback
+                return _shutdownSignal.Task;
+            }
 
-            return Task.CompletedTask;
+            NativeMethods.http_stop_incoming_requests();
+
+            try
+            {
+                // Wait for active requests to drain
+                if (_outstandingRequests > 0)
+                {
+                    RegisterCancelation();
+                }
+                else
+                {
+                    NativeMethods.http_stop_calls_into_mananged();
+                    _shutdownSignal.TrySetResult(null);
+                }
+            }
+            catch (Exception ex)
+            {
+                _shutdownSignal.TrySetException(ex);
+            }
+
+            return _shutdownSignal.Task;
         }
 
         public void Dispose()
         {
+            // This can get triggered too early...
+            _stopping = 1;
+            _shutdownSignal.TrySetResult(null);
```

Comment:
What about NativeMethods.http_stop_calls_into_mananged(); and http_stop_incoming_requests()?

Reply:
I can just set http_stop_incoming_requests().

Reply:
Why? Dispose may be called without StopAsync for error cases. What's to prevent requests in progress from AV'ing without stop_calls?

Reply:
Sorry I meant NativeMethods.http_stop_calls_into_managed(). 😨